src/supervisor/policyAnalysis.py

```diff
@@ -0,0 +1,194 @@
+        root_policy_number_trend_df.index= pandas.to_datetime(root_policy_number_trend_df.index, unit="s")
+        #node_cpu_trend_df =  node_cpu_trend_df.pivot( columns='node',values='value')
+        root_policy_number_trend_df.rename(columns={"value": "RootPolicyCount"}, inplace = True)
+        root_policy_number_trend_df.plot(title="Count of Root Policies",figsize=(30, 15))
+        plt.savefig('../../output/policy-root-count.png')
+        saveCSV(root_policy_number_trend_df,"policy-root-count",True)
+        plt.close('all') 
+
+    except Exception as e:
+        print(Fore.RED+"Error in getting number of root policies: ",e)
+        print(Style.RESET_ALL)    
+    print("=============================================")
+   
+    status=True
+    return status
+
+def getReplicatedPolicy(pc,startTime, endTime, step):
+    print("Get details of the replicated policies")
+
+    try:
+        propagated_policy = pc.custom_query('count(policy_governance_info{type="propagated"})')
+
+        propagated_policy_df = MetricSnapshotDataFrame(propagated_policy)
+        propagated_policy_df["value"]=propagated_policy_df["value"].astype(float)
+        propagated_policy_df.rename(columns={"value": "NumOfPropagatedPolicies"}, inplace = True)
+        print(propagated_policy_df[['NumOfPropagatedPolicies']].to_markdown())
+
+        
+        propagated_policy_number_trend = pc.custom_query_range(
+        query='count(policy_governance_info{type="propagated"})',
+            start_time=startTime,
+            end_time=endTime,
+            step=step,
+        )
+
+        propagated_policy_number_trend_df = MetricRangeDataFrame(propagated_policy_number_trend)
+        propagated_policy_number_trend_df["value"]=propagated_policy_number_trend_df["value"].astype(float)
+        propagated_policy_number_trend_df.index= pandas.to_datetime(propagated_policy_number_trend_df.index, unit="s")
+        #node_cpu_trend_df =  node_cpu_trend_df.pivot( columns='node',values='value')
+        propagated_policy_number_trend_df.rename(columns={"value": "PropagatedtPolicyTotal"}, inplace = True)
+        propagated_policy_number_trend_df.plot(title="Total Count of propagated Policies",figsize=(30, 15))
+        plt.savefig('../../output/policy-propagated-count.png')
+        saveCSV(propagated_policy_number_trend_df,"policy-propagated-count",True)
+        plt.close('all') 
+
+    except Exception as e:
+        print(Fore.RED+"Error in getting number of propagated policies: ",e)
+        print(Style.RESET_ALL)    
+    print("=============================================")
+   
+    status=True
+    return status
+
+
+def getReplicatedPolicyControllerResponseTime(pc,startTime, endTime, step):
+    print("Get Percentage of replicated-policy reconcile times le 1 sec")
+
+    # what is the issue with
+    # histogram_quantile(0.95, sum(rate(controller_runtime_reconcile_time_seconds_bucket{controller="replicated-policy"}[30m])))
```

Comment:
Ideally, I would have liked the 95th or 99th percentile controller reconcile time. While doing causal analysis, we could have seen how this varies with load on system which we are collecting thru other metrics. For some reason - possibly my lack of detail understanding of promql - I could not get the above PromQL to return data. I will try to resolve this.